# Comparing number of unique stable and version IDs mapping to UniProt primary accessions

11.18.19

mfpfox

### panel 2 figures


### design: 
- using xref ensembl files as cross-refrence source for ENSP ENST ENSG to uniprot primary ac
    - **new version 12.11.19 of Key not including ENST and ENSG, only matching at ENSP level**
    - creating stable ID KEY and filtering for only UKB IDs in 3963 set 
- merging xref file with fasta files from ensembl releases containing stable and version IDs, merge on key id (ENSP_ENST_ENSG)


### input files: 
**cross ref source**
- Homo_sapiens.GRCh37.85.uniprot.tsv
- Homo_sapiens.GRCh38.92.uniprot.tsv
- Homo_sapiens.GRCh38.94.uniprot.tsv
- Homo_sapiens.GRCh38.96.uniprot.tsv
- Homo_sapiens.GRCh38.97.uniprot.tsv

**fasta files with version and stable IDs**
- v85Homo_sapiens.GRCh37.pep.all.csv
- v92Homo_sapiens.GRCh38.pep.all.csv
- v94Homo_sapiens.GRCh38.pep.all.csv
- v96Homo_sapiens.GRCh38.pep.all.csv
- v97Homo_sapiens.GRCh38.pep.all.csv

**Uniprot IDs for filtering**
- UKBIDs_2018_3963.csv

# xref ref tsv files have cross-reference mapping to SOURCE == UniProt/RefSeq/PDB

##### VEP85  GRCh37 released 7.2016
##### VEP92 GRCh38 released 4.2018
##### VEP94 GRCh38  released 10.2018
##### VEP96 GRCh38  released 4.2019
##### VEP97 GRCh38 released 7.2019


#### ftp site for ensembl only has v85 cross ref file for GRCh37 mapping to uniprot
* Homo_sapiens.GRCh37.85.uniprot.tsv 
* why is this ensembl v85?? did they not have xref pipeline established at time of v75 release?
* ftp://ftp.ensembl.org/pub/grch37/release-96/tsv/homo_sapiens/
    
### FROM README_uniprot.tsv-

The files provided here are to give highlevel summaries of datasets in an easy to parse format. All TSV files are gzipped and their first line is a header line for file content.

+++++++++++++++++++++++++ Stable ID to uniprot Ac +++++++++++++++++++++++++

Provides mappings from Gene, Transcript and Translation stable identifiers to uniprot accessions with reports as to the % identity of the hit where applicable. **Dumps contain all Ensembl exeternal database names which started with UniProt so duplication of hits is possible.**

File are named Species.assembly.release.uniprot.tsv.gz

---

### ANOTHER SOURCE OF KEY ENSP--UNIPROT is UniProt ID mapping file, which includes uniprot isoform details and ENSG, ENST, and ENSP mapping to Ensembl (part of McGarvey paper)
> Not good idea to use in mismap of CYS LYS analysis becuase its isoform specific and the GOAL is to show stable ID failure

##### file made from UNIPROT ID mapping file (isoform uniprot style mapping to stable ensembl)
> isoCount.to_csv("UniprotIDMAPPING_isoformCount__labeledBefore_ENSPsource_3955unique entries.csv")
> ensp_labeled.to_csv("UniprotIDMAPPING_ENSPsourcetypefiltered_labeledBeforeEntry_7601")
> trueunilabeled.to_csv("UniprotIDMAPPING_trueunilabeled_incl_ENST_ENSP_19157.csv")

In [1]:
import os
import sys
import pandas as pd

from IPython.display import display, HTML
pd.set_option('display.max_columns', None)
pd.options.display.max_seq_items = 2000
display(HTML("<style>.container {width:90% !important;}</style>"))

sys.path.append("/Users/mariapalafox/Desktop/Toolbox")
from all_funx import *

os.chdir("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/TSV_UNIPROT_xref/MULTIMAPPING_IDs/")
print(os.listdir())

['ultimate_UKB_sequence_key_3953.csv', 'MISMAP2.0', 'GENE_ID_UKB.csv', 'GENE_ID_UKB.xlsx', '.DS_Store', 'ALL_releases_list_stdev_ID_versions_3953.csv', 'ENSEMBL_SEQ_v97_3953_UKBIDs.csv', 'ENSEMBL_SEQ_v96_3953_UKBIDs.csv', 'ALL_releases_list_stdev_ID_versions_3887.csv', 'ALL_releases_set_counts_ID_types_3953 copy.csv', 'SEQ_distance_UKBvsENSP', 'dict_3953_C_value_counts.csv', 'G6PD_pulledFrom_3953groupby_set_IDcount.csv', 'UKBID_3953set.csv', 'ALL_releases_set_counts_ID_types_3887.csv', 'dict_3953_CK_value_counts.csv', 'xref_merge_ENSEMBL_FASTA', 'uniprot-filtered-organism__Homo+sapiens+(Human)+[9606]_+AND+review--.tab', 'ENSEMBL_SEQ_v94_3953_UKBIDs.csv', 'GENE_ID_KEY_NOEXCEL_OPENING.csv', 'ALL_releases_list_stdev_ID_versions_3887 copy.csv', 'dict_3953_K_value_counts.csv', 'ENSEMBL_SEQ_v85_3953_UKBIDs.csv', 'dict_with_CK_counts_3953.csv', 'ALL_releases_set_counts_ID_types_3953.csv', 'ENSEMBL_SEQ_v92_3953_UKBIDs.csv', 'UKBID_3887set.csv']


# code snippets from 'Pmap_make_Ensembl_multimap_df.py'

In [5]:
# reference 2018 uniprot IDs
ID = pd.read_csv("UKBIDs_2018_3963.csv")
# making ID list to file
idls = list(ID.ID)
print(len(idls))

# ensembl cross-ref files for UKB stable --- ENSP ENST ENSG for releases
t85 = pd.read_csv("Homo_sapiens.GRCh37.85.uniprot.tsv", delimiter=r"\s+")
t92 = pd.read_csv("Homo_sapiens.GRCh38.92.uniprot.tsv", delimiter=r"\s+")
t94 = pd.read_csv("Homo_sapiens.GRCh38.94.uniprot.tsv", delimiter=r"\s+")
t96 = pd.read_csv("Homo_sapiens.GRCh38.96.uniprot.tsv", delimiter=r"\s+")
t97 = pd.read_csv("Homo_sapiens.GRCh38.97.uniprot.tsv", delimiter=r"\s+")

found85 = addcolumnconditionalDrop(idls, t85, 'xref', 'found_3963set')
found92 = addcolumnconditionalDrop(idls, t92, 'xref', 'found_3963set')
found94 = addcolumnconditionalDrop(idls, t94, 'xref', 'found_3963set')
found96 = addcolumnconditionalDrop(idls, t96, 'xref', 'found_3963set')
found97 = addcolumnconditionalDrop(idls, t97, 'xref', 'found_3963set')
print()
print("How many unique IDs from 3963 set in xref files?")
uniqueCount(found85, 'xref')
uniqueCount(found92, 'xref')
uniqueCount(found94, 'xref')
uniqueCount(found96, 'xref')
uniqueCount(found97, 'xref')

3963
  found_3963set   Count
0         False  196488
1          True   10194
dropping false
df shape post drop:  (10194, 10)
  found_3963set   Count
0         False  103325
1          True   10404
dropping false
df shape post drop:  (10404, 10)
  found_3963set   Count
0         False  104756
1          True   10627
dropping false
df shape post drop:  (10627, 10)
  found_3963set   Count
0         False  107283
1          True   10678
dropping false
df shape post drop:  (10678, 10)
  found_3963set   Count
0         False  106926
1          True   10579
dropping false
df shape post drop:  (10579, 10)

How many unique IDs from 3963 set in xref files?
len of col:  10194
len of col set:  3959

len of col:  10404
len of col set:  3959

len of col:  10627
len of col set:  3962

len of col:  10678
len of col set:  3962

len of col:  10579
len of col set:  3962



In [12]:
def diffList(L1, L2):
    s1 = set(L1)
    s2 = set(L2)
    diff = s1 - s2
    print("different number of IDs : ", len(diff))
    return list(diff)

In [13]:
# unique set of IDs in each release
uni85 = set(found85['xref'].tolist())
uni92 = set(found92['xref'].tolist())
uni94 = set(found94['xref'].tolist())
uni96 = set(found96['xref'].tolist())
uni97 = set(found97['xref'].tolist())

In [14]:
# difference between total 3963 IDs and number found in each release
d97 = diffList(idls,uni97)
d96 = diffList(idls,uni96)
d94 = diffList(idls,uni94)
d92 = diffList(idls,uni92)
d85 = diffList(idls,uni85)
print(d97)
print()
print(d96)
print()
print(d94)
print()
print(d92)
print()
print(d85)

different number of IDs :  1
different number of IDs :  1
different number of IDs :  1
different number of IDs :  4
different number of IDs :  4
['Q9HB19']

['Q9HB19']

['Q9HB19']

['Q9NVD7', 'P80303', 'P82921', 'Q9HB19']

['O75044', 'Q9Y2S7', 'Q96GY3', 'Q9NPI6']


| release | number of IDs found (n=3963) | IDs not found |
|---------|------------------------------|---------------|
| v85 | 3959 | 'O75044', 'Q9Y2S7', 'Q96GY3', 'Q9NPI6' |
| v92 | 3959 | 'Q9NVD7', 'P80303', 'P82921', 'Q9HB19' |
| v94 | 3962 | Q9HB19 |
| v96 | 3962 | Q9HB19 |
| v97 | 3962 | Q9HB19 |

# shared set of uniprot IDs in all ensembl releases?

In [15]:
# IDs in all releases
shared = uni85 & uni92 & uni94 & uni96 & uni97
print("set of shared IDs: ")
print(len(set(shared)))
shared = list(shared)
print(shared)

set of shared IDs: 
3955
['Q15459', 'Q9NVR2', 'O95373', 'P41223', 'Q6UN15', 'P25398', 'Q9NZB2', 'Q9BV86', 'Q96AT1', 'Q96E11', 'Q6P2Q9', 'Q9Y315', 'Q13951', 'Q8NFD5', 'Q9UHB9', 'Q9ULL5', 'O75334', 'Q13322', 'P39748', 'Q9NUI1', 'P11142', 'P13639', 'O14737', 'Q9NZ09', 'P11387', 'P13056', 'Q9Y508', 'Q6P2I3', 'Q9UHD9', 'Q8NC96', 'Q14683', 'Q92900', 'Q8TDH9', 'Q92688', 'P32929', 'Q5F1R6', 'P55809', 'Q7LDG7', 'Q08117', 'Q14C86', 'P28070', 'P26196', 'Q9GZL7', 'P11413', 'Q9H2K8', 'O75362', 'Q9Y4B4', 'Q96CX6', 'Q9H993', 'Q15061', 'P24844', 'Q9Y448', 'Q14181', 'O43572', 'P42575', 'O75964', 'P45880', 'Q9UMS0', 'Q6WCQ1', 'Q86UV5', 'P50402', 'O75935', 'P50897', 'Q8NEC7', 'Q92621', 'O43464', 'Q6ZVM7', 'Q7Z2T5', 'O60216', 'P62910', 'Q13422', 'Q7Z5J4', 'Q92609', 'P14550', 'Q9UN19', 'Q9Y312', 'A1L0T0', 'Q9BXB4', 'P30520', 'Q15048', 'Q9BVL4', 'Q70E73', 'Q9Y6W5', 'Q9NZ45', 'Q969Q0', 'Q15532', 'O15446', 'P30154', 'Q6GMV3', 'Q99986', 'P11021', 'P13716', 'P63279', 'Q9P265', 'P07910', 'Q92466', 'Q96JB5', 'O00

In [18]:
# drops rows where ID does not match list IDs 'Lever'
def dropNotLabeled(df, Lever):
    df['SharedUKB_IDs'] = np.where(df['xref'].isin(Lever),"True","False")
    df2 = df[df['SharedUKB_IDs']=="True"].copy()
    print("sorting ID column and resetting index: ")
    df2.sort_values(by=['xref'],inplace=True)
    df2.reset_index(drop=True, inplace=True)
    print("shape final cleaned df: ", df2.shape)
    return df2

In [19]:
shared85 = dropNotLabeled(found85,shared)
shared92 = dropNotLabeled(found92,shared)
shared94 = dropNotLabeled(found94,shared)
shared96 = dropNotLabeled(found96,shared)
shared97 = dropNotLabeled(found97,shared)

sorting ID column and resetting index: 
shape final cleaned df:  (10188, 11)
sorting ID column and resetting index: 
shape final cleaned df:  (10399, 11)
sorting ID column and resetting index: 
shape final cleaned df:  (10616, 11)
sorting ID column and resetting index: 
shape final cleaned df:  (10667, 11)
sorting ID column and resetting index: 
shape final cleaned df:  (10568, 11)


In [20]:
uniqueCount(shared85, 'xref')
uniqueCount(shared92, 'xref')
uniqueCount(shared94, 'xref')
uniqueCount(shared96, 'xref')
uniqueCount(shared97, 'xref')

len of col:  10188
len of col set:  3955

len of col:  10399
len of col set:  3955

len of col:  10616
len of col set:  3955

len of col:  10667
len of col set:  3955

len of col:  10568
len of col set:  3955



# creating stable ID key for each xref
- check that all are unique, drop redundant keys
- used to annotate ukb ID to ensembl fasta

In [24]:
def stableID_key(df, option):
    if option == 'xref':
        df['stableID_key'] = df['gene_stable_id'].astype(str) + \
'_' + df['transcript_stable_id'].astype(str) + \
'_' + df['protein_stable_id'].astype(str)
        
    if option == 'fasta':
        df['stableID_key'] = df['ENSG'].astype(str) + \
'_' + df['ENST'].astype(str) + \
'_' + df['ENSP'].astype(str)

In [25]:
stableID_key(shared85)
stableID_key(shared92)
stableID_key(shared94)
stableID_key(shared96)
stableID_key(shared97)

In [26]:
shared97.head(3)

gene_stable_id transcript_stable_id protein_stable_id    xref  \
0  ENSG00000033178      ENST00000420827   ENSP00000399234  A0AVT1   
1  ENSG00000033178      ENST00000322244   ENSP00000313454  A0AVT1   
2  ENSG00000117868      ENST00000251527   ENSP00000251527  A0FGR8   

             db_name info_type source_identity xref_identity linkage_type  \
0  Uniprot/SWISSPROT    DIRECT               -             -            -   
1  Uniprot/SWISSPROT    DIRECT             100           100            -   
2  Uniprot/SWISSPROT    DIRECT               -             -            -   

  found_3963set SharedUKB_IDs                                     stableID_key  
0          True          True  ENSG00000033178_ENST00000420827_ENSP00000399234  
1          True          True  ENSG00000033178_ENST00000322244_ENSP00000313454  
2          True          True  ENSG00000117868_ENST00000251527_ENSP00000251527

In [27]:
uniqueCount(shared85, 'stableID_key')
uniqueCount(shared92, 'stableID_key')
uniqueCount(shared94, 'stableID_key')
uniqueCount(shared96, 'stableID_key')
uniqueCount(shared97, 'stableID_key')

len of col:  10188
len of col set:  10188

len of col:  10399
len of col set:  10399

len of col:  10616
len of col set:  10616

len of col:  10667
len of col set:  10667

len of col:  10568
len of col set:  10568



**all xref files have unique ENSG_ENST_ENSP values, ready to map to fasta files**

In [30]:
shared85.to_csv("xref_v85_shared3955_10188.csv", index=False)
shared92.to_csv("xref_v92_shared3955_10399.csv", index=False)
shared94.to_csv("xref_v94_shared3955_10616.csv", index=False)
shared96.to_csv("xref_v96_shared3955_10667.csv", index=False)
shared97.to_csv("xref_v97_shared3955_10568.csv", index=False)

In [35]:
# simplify xref files
shared85 = shared85[['xref', 'stableID_key']].copy()
shared92 = shared92[['xref', 'stableID_key']].copy()
shared94 = shared94[['xref', 'stableID_key']].copy()
shared96 = shared96[['xref', 'stableID_key']].copy()
shared97 = shared97[['xref', 'stableID_key']].copy()

# ready to merge xref Uniprot ID with fasta file containing stable and version IDs for ensembl releases

In [28]:
# fasta to csv files with stable ID added by script
v85 = pd.read_csv("v85Homo_sapiens.GRCh37.pep.all.csv")
v92 = pd.read_csv("v92Homo_sapiens.GRCh38.pep.all.csv")
v94 = pd.read_csv("v94Homo_sapiens.GRCh38.pep.all.csv")
v96 = pd.read_csv("v96Homo_sapiens.GRCh38.pep.all.csv")
v97 = pd.read_csv("v97Homo_sapiens.GRCh38.pep.all.csv")

In [33]:
# adding key to fasta files
stableID_key(v85, 'fasta')
stableID_key(v92, 'fasta')
stableID_key(v94, 'fasta')
stableID_key(v96, 'fasta')
stableID_key(v97, 'fasta')

In [34]:
v85.head(3)

ENSPv             ENSP              ENSTv             ENST  \
0  ENSP00000451042.1  ENSP00000451042  ENST00000415118.1  ENST00000415118   
1  ENSP00000451515.1  ENSP00000451515  ENST00000434970.2  ENST00000434970   
2  ENSP00000452494.1  ENSP00000452494  ENST00000448914.1  ENST00000448914   

               ENSGv             ENSG  Length proSequence  \
0  ENSG00000223997.1  ENSG00000223997       2          EI   
1  ENSG00000237235.2  ENSG00000237235       3         PSY   
2  ENSG00000228985.1  ENSG00000228985       4        TGGY   

                                      stableID_key  
0  ENSG00000223997_ENST00000415118_ENSP00000451042  
1  ENSG00000237235_ENST00000434970_ENSP00000451515  
2  ENSG00000228985_ENST00000448914_ENSP00000452494

In [36]:
shared85.head(4)

xref                                     stableID_key
0  A0AVT1  ENSG00000033178_ENST00000322244_ENSP00000313454
1  A0AVT1  ENSG00000033178_ENST00000420827_ENSP00000399234
2  A0FGR8  ENSG00000117868_ENST00000251527_ENSP00000251527
3  A0JNW5  ENSG00000111647_ENST00000279907_ENSP00000279907

In [37]:
# merge on stable id to append uniprot ID with ensembl stable + version IDs
mer85 = pd.merge(v85,shared85,how='inner',on=['stableID_key'])
mer92 = pd.merge(v92,shared92,how='inner',on=['stableID_key'])
mer94 = pd.merge(v94,shared94,how='inner',on=['stableID_key'])
mer96 = pd.merge(v96,shared96,how='inner',on=['stableID_key'])
mer97 = pd.merge(v97,shared97,how='inner',on=['stableID_key'])

In [39]:
uniqueCount(shared85, 'stableID_key')
uniqueCount(shared92, 'stableID_key')
uniqueCount(shared94, 'stableID_key')
uniqueCount(shared96, 'stableID_key')
uniqueCount(shared97, 'stableID_key')

len of col:  10188
len of col set:  10188

len of col:  10399
len of col set:  10399

len of col:  10616
len of col set:  10616

len of col:  10667
len of col set:  10667

len of col:  10568
len of col set:  10568



In [40]:
uniqueCount(mer85, 'xref')
uniqueCount(mer92, 'xref')
uniqueCount(mer94, 'xref')
uniqueCount(mer96, 'xref')
uniqueCount(mer97, 'xref')

len of col:  10188
len of col set:  3955

len of col:  10399
len of col set:  3955

len of col:  10616
len of col set:  3955

len of col:  10667
len of col set:  3955

len of col:  10568
len of col set:  3955



In [42]:
# all rows in merged files have row count == to unique ENSP and ENST
uniqueCount(mer85, 'ENST')
uniqueCount(mer92, 'ENST')
uniqueCount(mer94, 'ENST')
uniqueCount(mer96, 'ENST')
uniqueCount(mer97, 'ENST')

len of col:  10188
len of col set:  10188

len of col:  10399
len of col set:  10399

len of col:  10616
len of col set:  10616

len of col:  10667
len of col set:  10667

len of col:  10568
len of col set:  10568



In [43]:
uniqueCount(mer85, 'ENSG')
uniqueCount(mer92, 'ENSG')
uniqueCount(mer94, 'ENSG')
uniqueCount(mer96, 'ENSG')
uniqueCount(mer97, 'ENSG')

len of col:  10188
len of col set:  4346

len of col:  10399
len of col set:  4334

len of col:  10616
len of col set:  4347

len of col:  10667
len of col set:  4338

len of col:  10568
len of col set:  4338



In [45]:
uniqueCount(mer85, 'ENSTv')
uniqueCount(mer92, 'ENSTv')
uniqueCount(mer94, 'ENSTv')
uniqueCount(mer96, 'ENSTv')
uniqueCount(mer97, 'ENSTv')

len of col:  10188
len of col set:  10188

len of col:  10399
len of col set:  10399

len of col:  10616
len of col set:  10616

len of col:  10667
len of col set:  10667

len of col:  10568
len of col set:  10568



In [52]:
def sort_col(df2):
    df2.sort_values(by=['xref', 'stableID_key'],inplace=True)
    df2.reset_index(drop=True, inplace=True)
    return df2

In [53]:
mer85 = sort_col(mer85)
mer92 = sort_col(mer92)
mer94 = sort_col(mer94)
mer96 = sort_col(mer96)
mer97 = sort_col(mer97)

In [54]:
mer85.head(4)

ENSPv             ENSP              ENSTv             ENST  \
0  ENSP00000313454.4  ENSP00000313454  ENST00000322244.5  ENST00000322244   
1  ENSP00000399234.2  ENSP00000399234  ENST00000420827.2  ENST00000420827   
2  ENSP00000251527.5  ENSP00000251527  ENST00000251527.5  ENST00000251527   
3  ENSP00000279907.7  ENSP00000279907  ENST00000279907.7  ENST00000279907   

                ENSGv             ENSG  Length  \
0   ENSG00000033178.8  ENSG00000033178    1052   
1   ENSG00000033178.8  ENSG00000033178     389   
2  ENSG00000117868.11  ENSG00000117868     893   
3   ENSG00000111647.8  ENSG00000111647    1464   

                                         proSequence  \
0  MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...   
1  MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...   
2  MTPPSRAEAGVRRSRVPSEGRWRGAEPPGISASTQPASAGRAARHC...   
3  MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...   

                                      stableID_key    xref  
0  ENSG00000033178_ENST00000322244_ENSP00000313454  A0AVT1  
1  ENSG00000033178_ENST00000420827_ENSP00000399234  A0AVT1  
2  ENSG00000117868_ENST00000251527_ENSP00000251527  A0FGR8  
3  ENSG00000111647_ENST00000279907_ENSP00000279907  A0JNW5

In [55]:
mer97.head(4)

ENSPv             ENSP               ENSTv             ENST  \
0  ENSP00000313454.4  ENSP00000313454  ENST00000322244.10  ENST00000322244   
1  ENSP00000399234.2  ENSP00000399234   ENST00000420827.2  ENST00000420827   
2  ENSP00000251527.6  ENSP00000251527  ENST00000251527.10  ENST00000251527   
3  ENSP00000279907.7  ENSP00000279907  ENST00000279907.12  ENST00000279907   

                ENSGv             ENSG  Length  \
0  ENSG00000033178.13  ENSG00000033178    1052   
1  ENSG00000033178.13  ENSG00000033178     389   
2  ENSG00000117868.16  ENSG00000117868     845   
3  ENSG00000111647.13  ENSG00000111647    1464   

                                         proSequence  \
0  MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...   
1  MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...   
2  MSGARGEGPEAGAGGAGGRAAPENPGGVLSVELPGLLAQLARSFAL...   
3  MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...   

                                      stableID_key    xref  
0  ENSG00000033178_ENST00000322244_ENSP00000313454  A0AVT1  
1  ENSG00000033178_ENST00000420827_ENSP00000399234  A0AVT1  
2  ENSG00000117868_ENST00000251527_ENSP00000251527  A0FGR8  
3  ENSG00000111647_ENST00000279907_ENSP00000279907  A0JNW5

In [56]:
mer85.to_csv("v85_fasta_merge_xref_3955IDs_10188.csv", index=False)
mer92.to_csv("v92_fasta_merge_xref_3955IDs_10399.csv", index=False)
mer94.to_csv("v94_fasta_merge_xref_3955IDs_10616.csv", index=False)
mer96.to_csv("v96_fasta_merge_xref_3955IDs_10667.csv", index=False)
mer97.to_csv("v97_fasta_merge_xref_3955IDs_10568.csv", index=False)

---
---
---
---
---
---
---
---
---

# creating set of ensembl files that contain only shared stableID_keys

## KEY ID version
1. ENSG_ENST_ENSP shared (n=8861 rows)
2. ENSP shared only 

### controlling for ENSP stable match, panel 2 figure ID version increments will use these files


In [8]:
# drops rows where ID does not match list IDs 'Lever'
def dropNotLabeled(df, sharedls, colname):
    df['Shared_IDs'] = np.where(df[colname].isin(sharedls), "True", "False")
    df2 = df[df['Shared_IDs'] == "True"].copy()
    print("sorting ID column and reseting index: ")
    df2.sort_values(by=[colname], inplace=True)
    df2.reset_index(drop=True, inplace=True)
    print("shape final cleaned df: ", df2.shape)
    return df2

In [9]:
def shared_IDs(df1, df2, df3, df4, df5, colname):
    # SHARED SET OF IDS
    # unique set of IDs in each release
    uni85 = set(df1[colname].tolist())
    uni92 = set(df2[colname].tolist())
    uni94 = set(df3[colname].tolist())
    uni96 = set(df4[colname].tolist())
    uni97 = set(df5[colname].tolist())
    # IDs in all releases
    shared = uni85 & uni92 & uni94 & uni96 & uni97
    print("set of shared IDs: ")
    print(len(set(shared)))
    shared = list(shared)
    # filtering release tsv files for only shared uniprot IDs
    shared85 = dropNotLabeled(df1, shared, colname)
    shared92 = dropNotLabeled(df2, shared, colname)
    shared94 = dropNotLabeled(df3, shared, colname)
    shared96 = dropNotLabeled(df4, shared, colname)
    shared97 = dropNotLabeled(df5, shared, colname)
    uniqueCount(shared85, colname)
    uniqueCount(shared92, colname)
    uniqueCount(shared94, colname)
    uniqueCount(shared96, colname)
    uniqueCount(shared97, colname)
    return shared85, shared92, shared94, shared96, shared97

## version 2 of key ID with ENSP only shared between all Ensembl releases
- 12.11.19
- version 1 done previously below

In [10]:
os.chdir("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/TSV_UNIPROT_xref/MULTIMAPPING_IDS/xref_merge_ENSEMBL_FASTA/")
print(os.listdir())


key85 = pd.read_csv("v85_fasta_merge_xref_3953IDs_10183.csv")
key92 = pd.read_csv("v92_fasta_merge_xref_3953IDs_10395.csv")
key94 = pd.read_csv("v94_fasta_merge_xref_3953IDs_10612.csv")
key96 = pd.read_csv("v96_fasta_merge_xref_3953IDs_10663.csv")
key97 = pd.read_csv("v97_fasta_merge_xref_3953IDs_10564.csv")

['v96_fasta_merge_xref_3953IDs_10663.csv', 'xref_v94_shared3955_10616.csv', 'v92_fasta_merge_xref_8861IDs_3887.csv', 'v92_fasta_merge_xref_3953IDs_10395.csv', 'xref_v97_shared3955_10568.csv', 'v97_fasta_merge_xref_3953IDs_10564.csv', 'v97_fasta_merge_xref_8861IDs_3887.csv', 'xref_v96_shared3955_10667.csv', 'v96_fasta_merge_xref_8861IDs_3887.csv', 'v94_fasta_merge_xref_8861IDs_3887.csv', 'v85_fasta_merge_xref_8861IDs_3887.csv', 'xref_v92_shared3955_10399.csv', 'v85_fasta_merge_xref_3953IDs_10183.csv', 'v94_fasta_merge_xref_3953IDs_10612.csv', 'xref_v85_shared3955_10188.csv']


In [7]:
key85.head(5)

ENSPv             ENSP              ENSTv             ENST  \
0  ENSP00000313454.4  ENSP00000313454  ENST00000322244.5  ENST00000322244   
1  ENSP00000399234.2  ENSP00000399234  ENST00000420827.2  ENST00000420827   
2  ENSP00000251527.5  ENSP00000251527  ENST00000251527.5  ENST00000251527   
3  ENSP00000279907.7  ENSP00000279907  ENST00000279907.7  ENST00000279907   
4  ENSP00000349285.3  ENSP00000349285  ENST00000356828.3  ENST00000356828   

                ENSGv             ENSG  Length  \
0   ENSG00000033178.8  ENSG00000033178    1052   
1   ENSG00000033178.8  ENSG00000033178     389   
2  ENSG00000117868.11  ENSG00000117868     893   
3   ENSG00000111647.8  ENSG00000111647    1464   
4   ENSG00000111647.8  ENSG00000111647     522   

                                         proSequence  \
0  MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...   
1  MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...   
2  MTPPSRAEAGVRRSRVPSEGRWRGAEPPGISASTQPASAGRAARHC...   
3  MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...   
4  MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...   

                                      stableID_key    xref  pro_ver  tx_ver  \
0  ENSG00000033178_ENST00000322244_ENSP00000313454  A0AVT1        4       5   
1  ENSG00000033178_ENST00000420827_ENSP00000399234  A0AVT1        2       2   
2  ENSG00000117868_ENST00000251527_ENSP00000251527  A0FGR8        5       5   
3  ENSG00000111647_ENST00000279907_ENSP00000279907  A0JNW5        7       7   
4  ENSG00000111647_ENST00000356828_ENSP00000349285  A0JNW5        3       3   

   gen_ver  
0        8  
1        8  
2       11  
3        8  
4        8

In [11]:
# instead of col as 'stableID_key' which was used in version 1, new col is ENSP only
m85, m92, m94, m96, m97 = shared_IDs(key85, key92, key94, key96, key97, 'ENSP')

set of shared IDs: 
8870
sorting ID column and reseting index: 
shape final cleaned df:  (8870, 14)
sorting ID column and reseting index: 
shape final cleaned df:  (8870, 14)
sorting ID column and reseting index: 
shape final cleaned df:  (8870, 14)
sorting ID column and reseting index: 
shape final cleaned df:  (8870, 14)
sorting ID column and reseting index: 
shape final cleaned df:  (8870, 14)
len of col:  8870
len of col set:  8870

len of col:  8870
len of col set:  8870

len of col:  8870
len of col set:  8870

len of col:  8870
len of col set:  8870

len of col:  8870
len of col set:  8870



In [12]:
uniqueCount(m85, 'xref')
uniqueCount(m92, 'xref')
uniqueCount(m94, 'xref')
uniqueCount(m96, 'xref')
uniqueCount(m97, 'xref')

len of col:  8870
len of col set:  3890

len of col:  8870
len of col set:  3890

len of col:  8870
len of col set:  3890

len of col:  8870
len of col set:  3890

len of col:  8870
len of col set:  3890



#  version 2 of key only includes 5 more ENSEMBL IDs (8870 vs version 1 8865) and uniprot IDs only increased by 1 (3890 vs version 1 3889 IDs)
- does not make a difference for figures, rather not save

## what was added uniprot IDs with different stable gene IDs?


In [ ]:
# m85.to_csv("v85_fasta_merge_xref_8865IDs_3889xref.csv", index=False)
# m92.to_csv("v92_fasta_merge_xref_8865IDs_3889xref.csv", index=False)
# m94.to_csv("v94_fasta_merge_xref_8865IDs_3889xref.csv", index=False)
# m96.to_csv("v96_fasta_merge_xref_8865IDs_3889xref.csv", index=False)
# m97.to_csv("v97_fasta_merge_xref_8865IDs_3889xref.csv", index=False)

---


## version 1 of key ID with ENSG_ENST_ENSP  shared between all Ensembl releases
- previously IDs have 3955, I filtered down to 3953 later, so these counts are slightly off

In [11]:
m85, m92, m94, m96, m97 = shared_IDs(mer85, mer92, mer94, mer96, mer97, 'stableID_key')
# only merge files with 3955 instead of 3953 corrected later (new files with 3953 used above for version 2 of key ID)

set of shared IDs: 
8865
sorting ID column and reseting index: 
shape final cleaned df:  (8865, 11)
sorting ID column and reseting index: 
shape final cleaned df:  (8865, 11)
sorting ID column and reseting index: 
shape final cleaned df:  (8865, 11)
sorting ID column and reseting index: 
shape final cleaned df:  (8865, 11)
sorting ID column and reseting index: 
shape final cleaned df:  (8865, 11)
len of col:  8865
len of col set:  8865

len of col:  8865
len of col set:  8865

len of col:  8865
len of col set:  8865

len of col:  8865
len of col set:  8865

len of col:  8865
len of col set:  8865



In [13]:
m85.head(10)

ENSPv             ENSP              ENSTv             ENST  \
0  ENSP00000363937.5  ENSP00000363937  ENST00000374804.5  ENST00000374804   
1  ENSP00000363940.5  ENSP00000363940  ENST00000374807.5  ENST00000374807   
2  ENSP00000363944.3  ENSP00000363944  ENST00000374811.3  ENST00000374811   
3  ENSP00000473471.1  ENSP00000473471  ENST00000484069.1  ENST00000484069   
4  ENSP00000344668.2  ENSP00000344668  ENST00000340022.2  ENST00000340022   
5  ENSP00000378011.2  ENSP00000378011  ENST00000394503.2  ENST00000394503   
6  ENSP00000378013.2  ENSP00000378013  ENST00000394505.2  ENST00000394505   
7  ENSP00000378015.1  ENSP00000378015  ENST00000394507.1  ENST00000394507   
8  ENSP00000410909.2  ENSP00000410909  ENST00000412043.2  ENST00000412043   
9  ENSP00000309103.3  ENSP00000309103  ENST00000309032.3  ENST00000309032   

                ENSGv             ENSG  Length  \
0  ENSG00000001497.12  ENSG00000001497     675   
1  ENSG00000001497.12  ENSG00000001497     717   
2  ENSG00000001497.12  ENSG00000001497     734   
3  ENSG00000001497.12  ENSG00000001497     291   
4  ENSG00000001631.10  ENSG00000001631     736   
5  ENSG00000001631.10  ENSG00000001631     688   
6  ENSG00000001631.10  ENSG00000001631     736   
7  ENSG00000001631.10  ENSG00000001631     736   
8  ENSG00000001631.10  ENSG00000001631     736   
9   ENSG00000002330.9  ENSG00000002330     168   

                                         proSequence  \
0  MSWESGAGPGLGSQGMDLVWSAWYGKCVKGKGSLPLSAHGIVVAWL...   
1  MSWESGAGPGLGSQGMDLVWSAWYGKCVKGKGSLPLSAHGIVVAWL...   
2  MSWESGAGPGLGSQGMDLVWSAWYGKCVKGKGSLPLSAHGIVVAWL...   
3  MSWESGAGPGLGSQGMDLVWSAWYGKCVKGKGSLPLSAHGIVVAWL...   
4  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
5  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
6  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
7  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
8  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
9  MFQIPEFEPSEQEDSSSAERGLGPSPAGDGPSGSGKHHRQAPGLLW...   

                                      stableID_key    xref Shared_IDs  
0  ENSG00000001497_ENST00000374804_ENSP00000363937  Q9Y4W2       True  
1  ENSG00000001497_ENST00000374807_ENSP00000363940  Q9Y4W2       True  
2  ENSG00000001497_ENST00000374811_ENSP00000363944  Q9Y4W2       True  
3  ENSG00000001497_ENST00000484069_ENSP00000473471  Q9Y4W2       True  
4  ENSG00000001631_ENST00000340022_ENSP00000344668  O00522       True  
5  ENSG00000001631_ENST00000394503_ENSP00000378011  O00522       True  
6  ENSG00000001631_ENST00000394505_ENSP00000378013  O00522       True  
7  ENSG00000001631_ENST00000394507_ENSP00000378015  O00522       True  
8  ENSG00000001631_ENST00000412043_ENSP00000410909  O00522       True  
9  ENSG00000002330_ENST00000309032_ENSP00000309103  Q92934       True

In [14]:
m97.head(10)

ENSPv             ENSP              ENSTv             ENST  \
0  ENSP00000363937.5  ENSP00000363937  ENST00000374804.9  ENST00000374804   
1  ENSP00000363940.5  ENSP00000363940  ENST00000374807.9  ENST00000374807   
2  ENSP00000363944.3  ENSP00000363944  ENST00000374811.7  ENST00000374811   
3  ENSP00000473471.1  ENSP00000473471  ENST00000484069.1  ENST00000484069   
4  ENSP00000344668.2  ENSP00000344668  ENST00000340022.6  ENST00000340022   
5  ENSP00000378011.2  ENSP00000378011  ENST00000394503.6  ENST00000394503   
6  ENSP00000378013.2  ENSP00000378013  ENST00000394505.6  ENST00000394505   
7  ENSP00000378015.1  ENSP00000378015  ENST00000394507.5  ENST00000394507   
8  ENSP00000410909.2  ENSP00000410909  ENST00000412043.6  ENST00000412043   
9  ENSP00000309103.3  ENSP00000309103  ENST00000309032.7  ENST00000309032   

                ENSGv             ENSG  Length  \
0  ENSG00000001497.16  ENSG00000001497     675   
1  ENSG00000001497.16  ENSG00000001497     717   
2  ENSG00000001497.16  ENSG00000001497     734   
3  ENSG00000001497.16  ENSG00000001497     291   
4  ENSG00000001631.15  ENSG00000001631     736   
5  ENSG00000001631.15  ENSG00000001631     688   
6  ENSG00000001631.15  ENSG00000001631     736   
7  ENSG00000001631.15  ENSG00000001631     736   
8  ENSG00000001631.15  ENSG00000001631     736   
9  ENSG00000002330.13  ENSG00000002330     168   

                                         proSequence  \
0  MSWESGAGPGLGSQGMDLVWSAWYGKCVKGKGSLPLSAHGIVVAWL...   
1  MSWESGAGPGLGSQGMDLVWSAWYGKCVKGKGSLPLSAHGIVVAWL...   
2  MSWESGAGPGLGSQGMDLVWSAWYGKCVKGKGSLPLSAHGIVVAWL...   
3  MSWESGAGPGLGSQGMDLVWSAWYGKCVKGKGSLPLSAHGIVVAWL...   
4  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
5  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
6  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
7  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
8  MGNPENIEDAYVAVIRPKNTASLNSREYRAKSYEILLHEVPIEGQK...   
9  MFQIPEFEPSEQEDSSSAERGLGPSPAGDGPSGSGKHHRQAPGLLW...   

                                      stableID_key    xref Shared_IDs  
0  ENSG00000001497_ENST00000374804_ENSP00000363937  Q9Y4W2       True  
1  ENSG00000001497_ENST00000374807_ENSP00000363940  Q9Y4W2       True  
2  ENSG00000001497_ENST00000374811_ENSP00000363944  Q9Y4W2       True  
3  ENSG00000001497_ENST00000484069_ENSP00000473471  Q9Y4W2       True  
4  ENSG00000001631_ENST00000340022_ENSP00000344668  O00522       True  
5  ENSG00000001631_ENST00000394503_ENSP00000378011  O00522       True  
6  ENSG00000001631_ENST00000394505_ENSP00000378013  O00522       True  
7  ENSG00000001631_ENST00000394507_ENSP00000378015  O00522       True  
8  ENSG00000001631_ENST00000412043_ENSP00000410909  O00522       True  
9  ENSG00000002330_ENST00000309032_ENSP00000309103  Q92934       True

In [15]:
uniqueCount(m85, 'xref')
uniqueCount(m92, 'xref')
uniqueCount(m94, 'xref')
uniqueCount(m96, 'xref')
uniqueCount(m97, 'xref')

len of col:  8865
len of col set:  3889

len of col:  8865
len of col set:  3889

len of col:  8865
len of col set:  3889

len of col:  8865
len of col set:  3889

len of col:  8865
len of col set:  3889



In [16]:
m85.to_csv("v85_fasta_merge_xref_8865IDs_3889xref.csv", index=False)
m92.to_csv("v92_fasta_merge_xref_8865IDs_3889xref.csv", index=False)
m94.to_csv("v94_fasta_merge_xref_8865IDs_3889xref.csv", index=False)
m96.to_csv("v96_fasta_merge_xref_8865IDs_3889xref.csv", index=False)
m97.to_csv("v97_fasta_merge_xref_8865IDs_3889xref.csv", index=False)